rust/src/eea_api.rs

```diff
@@ -0,0 +1,141 @@
+/*
+    The EEA WASM API functions
+        as outlined in the docs
+
+    These functions are defined in native code and imported into the EEA WASM module
+    Note: although WASM only really supports the data types i32, i64, f32, and f64
+          we can declare many of them different in the args for casting 
+*/
+use crate::configs::{
+    CONFIGS, Env, PTR_BUFFER_MESSAGE_TOPIC, PTR_BUFFER_MESSAGE_PAYLOAD,
+    BUFFER_MESSAGE_TOPIC_LENGTH, BUFFER_MESSAGE_PAYLOAD_LENGTH, TRACE_LEVELS
+};
+
+use rumqttc::QoS;
+use wasmer::{MemoryView, WasmPtr, Array};
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use std::sync::atomic::Ordering;
+use std::fs::{read, write};
+use std::thread;
+
+// print WASM info and error messages
+pub fn eea_trace(env: &Env, message: WasmPtr<u8, Array>, len: u32, level: i32) -> i32 {
+    // Use helper method on `WasmPtr` to read a utf8 string
+    let string = message.get_utf8_string(&env.memory, len).unwrap();
+    println!("{}: {}", TRACE_LEVELS[level as usize], string);
+
+    0
+}
+
+// publish mqtt message, the driver will queue it
+pub fn eea_send_message(
+    env: &Env,
+    topic: WasmPtr<u8, Array>,
+    topic_len: u32,
+    payload: WasmPtr<u8, Array>,
+    payload_len: u32, qos: i32
+) -> i32 {
+    let topic = topic.get_utf8_string(&env.memory, topic_len).unwrap();
+    let payload = payload.get_utf8_string(&env.memory, payload_len).unwrap();
+
+    let qos_type = match qos {
+        0 => QoS::AtMostOnce,
+        1 => QoS::AtLeastOnce,
+        _ => QoS::ExactlyOnce
+    };
+
+    env.mqtt_client.clone().publish(
```

Comment:
This should either be queued and processed in another thread or wrapped in a `task::spawn` (or something else so that this blocking publish does not stop the workflow). I noticed when running a 1-second timer that there are several long "pauses" in the script. My guess is those pauses are here because MQTT has blocked the workflow from running.

Reply:
Example from Python example - does not directly publish message. It queues message to be processed by the MQTT thread. This allow has the added benefit or queuing some amount of offline messages.

https://github.com/Losant/eea-examples/blob/main/python/eea_api.py#L20

ESP32 follows similar example, but has a max number of messages it'll queue for reasonable memory consumption.

https://github.com/Losant/eea-examples/blob/main/esp32/main/eea_api.cpp#L82

Reply:
hmm, didn't hit that issue but see how it could happen. I'll add a new queue and handle in the mqtt thread.